Added support for the user data multiLineShort

paulondc opened the pull request:
Adds support for the user data "multiLineShort" which works in conjunction with the user data "multiLine" that makes the StringParameterValueWidget to have a fixed short size which is used for brief descriptions.

#### Changed Log:
- GafferCortexUI: Added support for the user data multiLineShort in the StringParameterValueWidget

johnhaddon commented:
Hi Paulo, thanks for this. It seems like a really handy feature, but I think we could implement it in a way which is more useful elsewhere. I'm trying to get the UI to the point where everything is controlled by metadata on the plugs, so that every UI feature will be controllable from the UIEditor in the end. The aim is to have custom Boxes be first class citizens alongside real nodes.

Do you think we could implement it this way?

- Add methods to MultiLineTextWidget to specify a fixed height (in number of lines, rather than pixels). This would be a bit like TextWidget.setFixedCharacterWidth().
- Add support for a metadata entry in MultiLineStringPlugValueWidget, using it to control the size of the text widget.
- Map your parameter user data to the metadata entry. Maybe just allowing the user to specify the number of lines explicitly, rather than just "short".

That way we get useful functionality at every layer of the framework, rather than just something very specific to GafferCortex. What do you think?

paulondc commented:
Hi John,
Sounds great for me to get it for the base multiline text widget,I'll try to get back to this pull request as soon as possible.

paulondc commented:
Hi John,

I've made the changes we have discussed, I'm not super sure about that "fixedLines" name (maybe there is a better name for it, any thoughts ?). Also, noticed that you have to defined that  'QWIDGETSIZE_MAX' macro in the "PlugLayout" class and I had to do that as well, maybe we should centralize it, what do you think about it ?

Cheers,
Paulo

johnhaddon commented on python/GafferCortexUI/StringParameterValueWidget.py:
```diff
@@ -55,7 +57,20 @@ def __init__( self, parameterHandler, **kw ) :
 
 		if multiLine :
+			
+			# by default the multi-line widget gets auto-expanded in the op interface which works nicely when
```
Gaffer's Metadata system is designed so that you can auto-translate metadata from other sources (OSL, Arnold and RSL shader parameters are the main reason) on the fly, rather than explicitly setting them like this. We're already using this system for other metadata on parameters - there are examples here :

https://github.com/ImageEngine/gaffer/blob/master/python/GafferCortexUI/ParameterisedHolderUI.py#L169

We should do the same for this metadata too...

paulondc replied:
cool!!! I'm still finding my way to use the plugs metadata, thanks for pointing me that

johnhaddon commented on python/GafferUI/MultiLineTextWidget.py:
```diff
@@ -319,9 +328,57 @@ def __drop( self, widget, event ) :
+	def showEvent ( self, event ):
+
+		self.__widgetFullyBuilt = True		
+		self.__computeHeight()
```
I don't have your Qt guru status, so correct me if I'm wrong, but I don't think this is recommended way for a widget to request a size, and it'll break if the style changes after the widget is shown. Isn't this exactly what the `sizeHint()` mechanism is for? If you take a look at TextWidget.py, you'll see that that's how we implement the fixed character width there, and it seems to work pretty well, as well as (I believe) being "the Qt way" :

https://github.com/ImageEngine/gaffer/blob/master/python/GafferUI/TextWidget.py#L386

What do you think?

paulondc replied:
Every time that Qt makes me to look through the cpp implementation to see what is really happening I think I don't know anything about Qt at all!!

The ListContainer in gaffer uses as size constraint "QLayout::SetMinAndMaxSize" which is driven by the widget minimumSize/maximumSize that is why I'm using setFixedHeight instead of sizeHint (setFixedHeight is basically a shortcut to set the same value for the minimum/maximum height).

So the sizeHint information for that constraint would not affect the limits of the widget
http://doc.qt.io/qt-4.8/qlayout.html#SizeConstraint-enum

Unless :D a custom sizePolicy gets set to the widget which is what is happening in the TextWidget, both ways work fine, but I gotta say now thinking more about it, changing the sizePolicy for the widget seems more sutiable for this case, there is just one thing I noticed that is missing in the TextWidget which is when the sizePolicy changes after the widget being displayed, so if you don't call the updateGeometry after that the widget will keep stuck displaying the geometry based in the old sizePolicy (Qt weirdness). Could I add that in the end of TextWidget.setFixedCharacterWidth ?

johnhaddon replied:
Ah yes, the joys of taking a stroll through the Qt source code :) It definitely seems necessary to read the source to have any chance of understanding what is going on half the time. On that subject, I just took a look at `QWidget::setSizePolicy()` and it already includes a call to `updateGeometry()`, so I'm not sure if the problem you mentioned in `TextWidget` will need some other fix. Perhaps you could start by making a unit test showing the problem? I've certainly no objection to fixing it, but it seems like we might need to understand it a little more first...

paulondc replied:
I've sent to you a test case by email, but looking at the QWidget.setSizePolicy implementation it only calls the updateGeometry when is a different size policy otherwise it returns right away (that thing is easy to get mislead), that explains why it is failing when setting a different character width for the same size policy, anyway I had to look through the Qt code again :D

johnhaddon replied:
Ah - that makes sense. Thanks for the explanation.

johnhaddon commented on python/GafferUI/MultiLineTextWidget.py:
```diff
@@ -132,6 +133,14 @@ def getWrapMode( self ) :
 			QtGui.QTextOption.WrapAtWordBoundaryOrAnywhere : self.WrapMode.WordOrCharacter,
 		}[self._qtWidget().wordWrapMode()]
 
+	def setFixedLines( self, fixedLines ) :
```
How about we call this `setFixedLineHeight()` so that it matches `TextWidget.setFixedCharacterWidth()`?. To be honest, I think `setFixedCharacterWidth()` sounds a bit clunky, but I'd rather they both sounded clunky in the same way - it makes it a bit easier to remember.

paulondc replied:
Turns out that is really hard to figure out these names, but I gotta say that I prefer your suggestion

johnhaddon commented:
Hi Paulo, I've added a few comments which hopefully you think are reasonable. Do you think you could address them and then rebase everything to have a nice clean history with one commit for the MultiLineTextWidget, one for the MultiLineStringPlugValueWidget, and one for the parameter UI? Then I think we'll have a nice tidy implementation of some very handy functionality...
Thanks...
John

paulondc commented:
Thanks for the comments John, I'm gonna address this pull request following your suggestions.

Cheers,
Paulo

paulondc commented:
Hi John,
I've addressed the changes and rebased it, thanks for the suggestions!
Could you take a look into it ?

Cheers

paulondc commented:
Hi John,
I've addressed the changes and rebased it, thanks for the suggestions!
Could you take a look into it ?

Cheers

johnhaddon commented:
Thanks Paulo - looks great apart from the two minor comments I made. Maybe you could address those and then we can merge?

paulondc commented:
Hi John,

I was not aware about widget.bound and widget.size where those added recently in gaffer ? they are quite handy :)

I've added tests for the TextWidget.fixedCharacterWidth and MultiLineTextWidget.fixedLineHeight. Also, fixed that plug metadata check which was happening outside of the plug test.

Cheers

johnhaddon commented:
Nice one - thanks Paulo. Looks like the `size()` and `bound()` methods have been around since 2011 and 2012 respectively, so not particularly new. As a general rule, if you find yourself needing direct access to the underlying QWidget, it's good to look for an alternative in the public API - the aim with GafferUI is to entirely shield the user from the underlying (Qt, but was once GTK) implementation. I'm sure there's the odd hole still, but I'd like to get them all patched up.
